src/tools/PublicAnalyzers/AvaloniaPropertyAnalyzer.cs

```diff
@@ -0,0 +1,463 @@
+        "Superfluous owner: {0} is already an owner of {1} via {2}",
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        "Ownership of an AvaloniaProperty is inherited along the type hierarchy. There is no need for a derived type to assert ownership over a base type's properties. This diagnostic can be a symptom of an incorrect property owner elsewhere.",
+        InappropriateReadWriteTag);
+
+    private static readonly DiagnosticDescriptor DuplicatePropertyName = new(
+        "AVP1020",
+        "AvaloniaProperty names should be unique within each class",
+        "Name collision: {0} has the same name as {1}",
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        "Querying for an AvaloniaProperty by name requires that each property associated with a type have a unique name.",
+        NameCollisionTag);
+
+    private static readonly DiagnosticDescriptor AmbiguousPropertyName = new(
+        "AVP1021",
+        "There should be an unambiguous relationship between the CLR properties and Avalonia properties of a class",
+        "Name collision: {0} owns multiple Avalonia properties with the name '{1}' {2}",
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        "It is unclear which AvaloniaProperty this CLR property refers to. Ensure that each AvaloniaProperty associated with a type has a unique name. If you need to change behaviour of a base property in your class, call its AddOwner method and provide new metadata.",
+        NameCollisionTag);
+
+    private static readonly DiagnosticDescriptor PropertyNameMismatch = new(
+        "AVP1022",
+        "An AvaloniaProperty object should be stored in a field or CLR property which reflects its name",
+        "Bad name: An AvaloniaProperty named '{0}' is being assigned to {1}. These names do not relate.",
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        "An AvaloniaProperty should be stored in a field or property which contains its name. For example, a property named \"Brush\" should be assigned to a field called \"BrushProperty\".\nPrivate symbols are exempt from this diagnostic.",
+        NameCollisionTag);
+
+    private static readonly DiagnosticDescriptor AccessorSideEffects = new(
+        "AVP1030",
+        "StyledProperty accessors should not have side effects",
+        "Side effects: '{0}' is an AvaloniaProperty which can be {1} without the use of this CLR property. This {2} accessor should do nothing except call {3}.",
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        "The AvaloniaObject.GetValue and AvaloniaObject.SetValue methods are public, and do not call any user CLR properties. To execute code before or after the property is set, create a Coerce method or a PropertyChanged subscriber.",
```

Comment:
Overriding OnPropertyChanged method is the recommended way. As PropertyChanged subscriber will add overhead (minor, though). And coercing is more about changing value during property changing and theoretically shouldn't have side effects too (practically, there is no problems with that though).

Reply:
The description mentions `OnPropertyChanged` now. I'm surprised that it's the recommended approach though. It's a virtual method which evaluates a multitude if statements every time ANY property changes on the object, while a class handler should only be executed if one single property changes on a small subset of types. This is not relevant for the current PR though.

I agree about coercion methods, but it will be much harder to define when they have a side effect. A reasonable starting point would be warning about calls to `SetValue` within one.

Reply:
For some reason I misread it as AvaloniaProperty.Changed observable.
AvaloniaObject.PropertyChanged is basically the same as OnPropertyChanged, except one is an event and another is a virtual method.

Reply:
Oh, that's a mistake on my part too! I did actually mean `AvaloniaProperty.Changed`.